allennlp/data/semparse/type_declarations/nlvr_type_declaration.py

```diff
@@ -0,0 +1,185 @@
+    arguments, and looks like ``filter(initial_set_of_boxes, attribute_function, target_attribute)`` .
+    Hence, the signature of the filter function is <b,<<b,#1>,<#1,b>>>.
+    """
+    @property
+    def _signature(self) -> str:
+        return "<b,<<b,#1>,<#1,b>>>"
+
+    @overrides
+    def resolve(self, other: Type) -> Optional[Type]:
+        if not isinstance(other, ComplexType):
+            return None
+        if not isinstance(other.second, ComplexType):
+            return None
+        if not isinstance(other.second.first, ComplexType) or not isinstance(other.second.second, ComplexType):
+            return None
+        resolved_type = other.resolve(ComplexType(BOX_TYPE, ComplexType(ComplexType(BOX_TYPE, ANY_TYPE),
+                                                                        ComplexType(ANY_TYPE, BOX_TYPE))))
+        if not resolved_type:
+            return None
+        first_placeholder = resolved_type.second.first.second
+        second_placeholder = resolved_type.second.second.first
+        resolved_placeholder = first_placeholder.resolve(second_placeholder)
+        if not resolved_placeholder:
+            return None
+        return BoxFilterType(BOX_TYPE, ComplexType(ComplexType(BOX_TYPE, resolved_placeholder),
+                                                   ComplexType(resolved_placeholder, BOX_TYPE)))
+
+    @overrides
+    def get_application_type(self, argument_type: Type) -> Type:
+        return self.second
+
+
+class AssertType(PlaceholderType):
+    """
+    This is the type of assert operations. Corresponding Python function looks like
+    ``assert(attribute, target_attribute)`` , where the attribute and target attribute can either be numbers,
+    shapes or colors, and the return type is boolean. So the signature of the function is <#1,<#1,t>>.
+    """
+    @property
+    def _signature(self) -> str:
+        return "<#1,<#1,t>>"
+
+    @overrides
+    def resolve(self, other: Type) -> Optional[Type]:
+        if not isinstance(other, ComplexType):
+            return None
+        if not isinstance(other.second, ComplexType):
+            return None
+        resolved_type = other.resolve(ComplexType(ANY_TYPE, ComplexType(ANY_TYPE, TRUTH_TYPE)))
+        if not resolved_type:
+            return None
+        resolved_placeholder = resolved_type.first.resolve(resolved_type.second.first)
+        if not resolved_placeholder:
+            return None
+        return AssertType(resolved_placeholder, ComplexType(resolved_type, TRUTH_TYPE))
+
+    @overrides
+    def get_application_type(self, argument_type: Type) -> Type:
+        return ComplexType(argument_type, TRUTH_TYPE)
+
+
+class CountType(PlaceholderType):
+    """
+    This is the type of the count function. Corresponding Python function looks like ``count(entities)`` ,
+    where entities are either objects or boxes. Type signature is <#1,e>.
```

Comment:
`e` is the `NUM_TYPE`?  Why?

Reply:
`e` is the entity type in NLTK and all constants default to it. I dealt with constants having different types by letting the world specify `constant_type_prefixes`: https://github.com/pdasigi/allennlp/blob/c04c985521ffad9ec318c75890b012d010336cc9/allennlp/data/semparse/worlds/world.py#L19-L21
This requires mapping names of constants to contain their types as prefixes as well. In NLVR, since numbers are the only constants we see in the logical forms, I let them default to entities to avoid having to translate them. I shall write a comment about this.